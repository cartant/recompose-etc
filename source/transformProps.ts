import * as React from "react";
import {
  componentFromStreamWithConfig,
  mapper,
  setDisplayName,
  Subscribable,
  wrapDisplayName
} from "recompose";
import { Observable } from "rxjs/Observable";
import { from } from "rxjs/observable/from";
import { map } from "rxjs/operators/map";
import { withLatestFrom } from "rxjs/operators/withLatestFrom";
import { rxjsObservableConfig } from "./rxjsObservableConfig";

export function transformProps<TOuterProps, TInnerProps>(
  transform: mapper<Observable<TOuterProps>, Observable<TInnerProps>>
): React.ComponentType<TInnerProps> & { ChildProps: TInnerProps } {
  const componentFromStream = componentFromStreamWithConfig(rxjsObservableConfig);
  const Component = componentFromStream<TOuterProps & { children?: (props: TInnerProps) => React.ReactNode }>(
    props$ => transform(from(props$)).pipe(
      withLatestFrom(props$),
      map(([innerProps, props]) => props.children!(innerProps))
    )
  );
  if (process.env.NODE_ENV !== "production") {
    return setDisplayName(wrapDisplayName(Component, "transformProps"))(Component) as any;
  }
  return Component as any;
}
