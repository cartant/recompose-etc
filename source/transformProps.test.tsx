import { shallow } from "enzyme";
import * as React from "react";
import * as renderer from "react-test-renderer";
import { Observable } from "rxjs";
import { mapTo } from "rxjs/operators/mapTo";
import { transformProps } from "./transformProps";

const Component = transformProps((props$: Observable<{ name: string }>) => props$.pipe(mapTo({ name: "bob" })));

it("should transform props using the children prop", () => {
  const wrapper = shallow(
    <Component
      children={({ name }) => <span>{name}</span>}
      name={"alice"}
    />
  );
  expect(wrapper.html()).toEqual(expect.stringMatching(/bob/));
});

it("should transform props using implicit children", () => {
  const wrapper = shallow(
    <Component name={"alice"}>
      {({ name }) => <span>{name}</span>}
    </Component>
  );
  expect(wrapper.html()).toEqual(expect.stringMatching(/bob/));
});

it("should render correctly using the children prop", () => {
  const rendering = renderer.create(
    <Component
      children={({ name }) => <span>{name}</span>}
      name={"alice"}
    />
  ).toJSON();
  expect(rendering).toMatchSnapshot();
});

it("should render correctly using implicit children", () => {
  const rendering = renderer.create(
    <Component name={"alice"}>
      {({ name }) => <span>{name}</span>}
    </Component>
  ).toJSON();
  expect(rendering).toMatchSnapshot();
});
