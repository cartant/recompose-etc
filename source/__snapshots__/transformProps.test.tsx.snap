// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly using implicit children 1`] = `
<span>
  bob
</span>
`;

exports[`should render correctly using the children prop 1`] = `
<span>
  bob
</span>
`;
